signalr-go-server/pkg/signalr/server.go

```diff
@@ -0,0 +1,208 @@
+	defaultHubClients defaultHubClients
+	groupManager GroupManager
+}
+
+func NewServer(hubPrototype HubInterface) *Server {
+	lifetimeManager := defaultHubLifetimeManager{}
+	return &Server{
+		hubPrototype:    hubPrototype,
+		lifetimeManager: &lifetimeManager,
+		defaultHubClients: defaultHubClients{
+			lifetimeManager: &lifetimeManager,
+			allCache:        allClientProxy{lifetimeManager: &lifetimeManager},
+		},
+		groupManager: &defaultGroupManager{
+			lifetimeManager: &lifetimeManager,
+		},
+	}
+}
+
+func (s *Server) MessageLoop(conn HubConnection, connectionID string, protocol HubProtocol) {
+	streamer := newStreamer(conn)
+	streamClient := newStreamClient()
+	hubInfo := s.newHubInfo(connectionID)
+	hubInfo.lifetimeManager.OnConnected(conn)
+	hubInfo.hub.OnConnected(connectionID)
+
+	for conn.IsConnected() {
+		if message, err := conn.Receive(); err != nil {
+			fmt.Println(err)
+			break
+		} else {
+			fmt.Printf("Message received %v\n", message)
+			switch message.(type) {
+			case invocationMessage:
+				invocation := message.(invocationMessage)
+				// Dispatch invocation here
+				if method, ok := hubInfo.methods[strings.ToLower(invocation.Target)]; !ok {
+					// Unable to find the method
+					conn.Completion(invocation.InvocationID, nil, fmt.Sprintf("Unknown method %s", invocation.Target))
+				} else if in, clientStreaming, err := buildMethodArguments(method, invocation, streamClient, protocol); err != nil {
+					// argument build failed
+					conn.Completion(invocation.InvocationID, nil, err.Error())
+				} else if clientStreaming {
+					// let the receiving method run idependently
+					go func() {
+						defer func() {
+							if err := recover(); err != nil {
+								conn.Completion(invocation.InvocationID, nil, fmt.Sprintf("%v\n%v", err, string(debug.Stack())))
+							}
+						}()
+						method.Call(in)
+					}()
+				} else {
+					result := func() []reflect.Value {
+						defer func() {
+							if err := recover(); err != nil {
+								conn.Completion(invocation.InvocationID, nil, fmt.Sprintf("%v\n%v", err, string(debug.Stack())))
+							}
+						}()
+						return method.Call(in)
+					}()
+					returnInvocationResult(conn, invocation, streamer, result)
+				}
+			case cancelInvocationMessage:
+				streamer.Stop(message.(cancelInvocationMessage).InvocationID)
+			case streamItemMessage:
+				streamClient.receiveStreamItem(message.(streamItemMessage))
+			case completionMessage:
+				streamClient.receiveCompletionItem(message.(completionMessage))
+			case hubMessage:
+				// Ping
+			}
+		}
+	}
+	hubInfo.hub.OnDisconnected(connectionID)
+	hubInfo.lifetimeManager.OnDisconnected(conn)
+}
+
+type hubInfo struct {
+	hub             HubInterface
+	lifetimeManager HubLifetimeManager
+	methods         map[string]reflect.Value
+}
+
+func (s *Server) newHubInfo(connectionID string) *hubInfo {
+
+	// Copy hubPrototype
+	hub := reflect.New(reflect.ValueOf(s.hubPrototype).Elem().Type()).Interface().(HubInterface)
```

Comment:
I would revert this for now.

Reply:
ok

Reply:
See [45c7f86](https://github.com/davidfowl/signalr-ports/pull/6/commits/45c7f86212198900a4488c10d444bfc019592f7b)

Reply:
@davidfowl I tried to find out how caller support is implemented in AspNetCore. DefaultHubDispatcher.OnConnectedAsync creates a new Hub instance and sets its Client property to a new HubCallerClients instance which carries the ConnectionId and implements Caller() the same way contextHubClients did here. 
Is there a difference in invocation time? I did not debug, but its name indicates that OnConnectedAsync might be called when the connection is established. Here, setting the clients property is set after the handshake is completed and various OnConnected funcs are called
Do you have a different pattern in mind?

Reply:
It works well in asp.net core because we also have passing state to hubs that are activated by the framework. I'm concerned with supporting per hub invocation without a solution so lets leave it for last.

Reply:
passing state > Are you thinking about Context.Items or dependency injection of hub c'tor parameters? Context.Items might be added to the HubContext interface. For DI, there seems to be no widely adopted DI container equal to AspNetCore DI. But what about passing a `func newHub() HubInterface` to `newServer(...)`?
BTW: When looking at the //Copy prototype code, I saw that I implemented a scoped Hub, not a transient one. If we decide to do changes here, the newHub() function needs to be called inside the messageLoops for-loop and getting the reflected method has to be moved there also.

Reply:
> dependency injection of hub c'tor parameters?

This.